test-suite/src/test/scala/org/scalajs/testsuite/niobuffer/BufferFactory.scala

```diff
@@ -0,0 +1,253 @@
+package org.scalajs.testsuite.niobuffer
+
+import scala.language.implicitConversions
+
+import scala.reflect._
+
+import java.nio._
+
+import scala.scalajs.js
+
+abstract class BufferFactory {
+  type BufferType <: Buffer with Comparable[BufferType]
+  type ElementType
+
+  implicit val elemClassTag: ClassTag[ElementType]
+
+  implicit def elemFromInt(value: Int): ElementType
+
+  implicit def elemToJSAny(elem: ElementType): js.Any
+
+  implicit def bufferAdapter(
+      buffer: BufferType): BufferAdapter[BufferType, ElementType]
+
+  val createsReadOnly: Boolean = false
+
+  def allocBuffer(capacity: Int): BufferType
+
+  def allocBuffer(pos: Int, limit: Int, capacity: Int): BufferType = {
+    val buf = allocBuffer(capacity)
+    buf.limit(limit).position(pos)
+    buf
+  }
+
+  def elemRange(start: Int, end: Int): Array[ElementType] =
+    (start until end).map(elemFromInt).toArray
+
+  def withContent(capacity: Int, content: ElementType*): BufferType =
+    withContent(0, capacity, capacity, content: _*)
+
+  def withContent(pos: Int, limit: Int, capacity: Int,
+      content: ElementType*): BufferType = {
+    val buf = allocBuffer(pos, limit, capacity)
+    buf.put(content.toArray)
+    buf.position(pos)
+    buf
+  }
+}
+
+object BufferFactory {
+  abstract class ByteBufferFactory extends BufferFactory {
```

Comment:
Hum, now trying to move them, makes me understand my feeling and why I put them here in the first place. Essentially they are the only types of factories, just like the `XBuffer` classes are the only types of `Buffer` (actually I should make `BufferFactory` sealed). So, as types, they are much more general than particular implementations of these factories. It can be quite normal for some method to require a factory of a given buffer type, whereas it makes no sense to require a particular implementation of these factories.

I'll keep them where they are for now. If you still want me to move them, I'll add a commit at the end (because moving them from this commit additionally causes endless conflicts).